### DISCUSSION TITLE
Is there a way to debug serialization issue in CoreWCF ? Similar to WCF Trace and svclog configuration

### DISCUSSION THREAD

**Discussion by MilanRaval** [CoreWCF/CoreWCF#1270] [Q&A] [answered]
While converting a WCF service to CoreWCF, existing method works fine and returns results but client is getting it as null value.

**Comment by mconnew** [comment]
What binding are you using? I suspect the issue is with one of the .NET runtime types that broke serialization as they've removed the ability to serialize some of them, e.g. IPAddress can no longer be serialized. If that's the case, then you'll need to create a surrogate for it.

**Reply by MilanRaval** [reply]
I am using NetTcp

**Reply by mconnew** [reply, marked as answer]
The easiest way is probably to create an IDispatchMessageInspector, and in BeforeSendReply, call Message.ToString(). That will generate a generic XML output of the reply. It won't be exactly the same as what's sent as NetTcp sends it in a binary format, and the envelope/addressing versions likely won't match, and there might be a couple of missing SOAP headers that get added later. But the SOAP body will be a sufficiently accurate representation that it will work. You could also do the same thing on the client side using an IClientMessageInspector and call ToString() on the reply message in AfterReceiveReply as that's before we deserialize the reply.  
It should be reasonably obvious what the problem is in the XML. Feel free to provide the XML (or partial XML where it's different). Please provide the details of the class that is being differently serialized too if possible.

**Reply by MilanRaval** [reply]
Thanks mconnew ! 

It helped to figure out the differences in the xmlns, the issue was we use ContractNameSpace attribute to map clr namespace with a common namespace, and that attribute was missing in some data contracts, fixing it fixed the issue.

**Reply by danreg** [reply]
@mconnew how can I add my implementation of IDispatchMessageInspector to corewcf so that it get's called? Any code need to be added to Startup.cs?

**Reply by MilanRaval** [reply]
Add ServiceBehavior, like below to the IApplicationBuilder app

```
 app.UseServiceModel(builder =>
 {
     builder.AddService<Service>(serviceOptions => { })
     .AddServiceEndpoint<Service, Contract>(netTcpBinding, url);
     builder.ConfigureServiceHostBase<Service>(serviceHost =>
     {
         serviceHost.Description.Behaviors.Add(new ServiceBehavior());
     });
 });

```
ServiceBehavior implements IServiceBehavior, add below code to ApplyDispatchBehavior method of the Service Behavior class, 

```
foreach (ChannelDispatcherBase cdb in serviceHostBase.ChannelDispatchers)
{
    ChannelDispatcher cd = cdb as ChannelDispatcher;
    if (cd != null)
    {
        foreach (EndpointDispatcher eDisc in cd.Endpoints)
        {
              eDisc.DispatchRuntime.MessageInspectors.Add(new MessageInspector());
        }
    }
}
```

**Reply by danreg** [reply]
@MilanRaval great thx! Is there some documentation I'm missing or where do you get all the information?
